osquery/tables/system/windows/battery.cpp

```diff
@@ -0,0 +1,264 @@
+/**
+ * Copyright (c) 2014-present, The osquery authors
+ *
+ * This source code is licensed as defined by the LICENSE file found in the
+ * root directory of this source tree.
+ *
+ * SPDX-License-Identifier: (Apache-2.0 OR GPL-2.0-only)
+ */
+
+#include <string>
+
+// clang-format off
+#include <windows.h>
+#include <ioapiset.h>
+#include <Batclass.h>
+#include <Poclass.h>
+#include <setupapi.h>
+#include <devguid.h>
+#include <devioctl.h>
+// clang-format on
+
+#include <osquery/logger/logger.h>
+#include <osquery/sql/sql.h>
+#include <osquery/utils/conversions/windows/strings.h>
+#include <osquery/utils/scope_guard.h>
+#include <osquery/utils/system/errno.h>
+
+namespace osquery {
+namespace tables {
+
+std::string batteryQueryInformationString(
+    HANDLE hBattery,
+    ULONG batteryTag,
+    BATTERY_QUERY_INFORMATION_LEVEL informationLevel) {
+  BATTERY_QUERY_INFORMATION bqi = {0};
+  bqi.InformationLevel = informationLevel;
+  bqi.BatteryTag = batteryTag;
+  // 1025 characters should be way more than enough for the values retrieved
+  // from this function. It shouldn't overflow anyway due to providing size in
+  // the DeviceIoControl call.
+  std::wstring resWstring(1025, L'\0');
+  DWORD resSize(0);
+
+  if (!DeviceIoControl(hBattery,
+                       IOCTL_BATTERY_QUERY_INFORMATION,
+                       &bqi,
+                       sizeof(bqi),
+                       resWstring.data(),
+                       static_cast<DWORD>(resWstring.size()) * sizeof(wchar_t),
+                       nullptr,
+                       nullptr)) {
+    if (ERROR_INVALID_FUNCTION == GetLastError()) {
+      LOG(INFO) << "Battery does not support information level "
+                << informationLevel;
+    } else {
+      LOG(ERROR) << "Failed to get battery information level "
+                 << informationLevel << ": code " << GetLastError();
+    }
+    return "";
+  }
+
+  return wstringToString(resWstring);
+}
+
+QueryData genBatteryInfo(QueryContext& context) {
+  QueryData results;
+
+  // Adapted from Microsoft example:
+  // https://learn.microsoft.com/en-us/windows/win32/power/enumerating-battery-devices
+  // Enumerate the batteries and ask each one for information.
+  HDEVINFO hdev = SetupDiGetClassDevs(
+      &GUID_DEVCLASS_BATTERY, 0, 0, DIGCF_PRESENT | DIGCF_DEVICEINTERFACE);
+  if (hdev == INVALID_HANDLE_VALUE) {
+    LOG(ERROR) << "Failed to initialize handle for enumerating batteries: "
+               << GetLastError();
+    return results;
+  }
+  auto const hdevGuard =
+      scope_guard::create([&]() { SetupDiDestroyDeviceInfoList(hdev); });
+
+  // Limit search to 100 batteries max
+  for (int idev = 0; idev < 100; idev++) {
+    SP_DEVICE_INTERFACE_DATA did = {0};
+    did.cbSize = sizeof(did);
+
+    if (!SetupDiEnumDeviceInterfaces(
+            hdev, 0, &GUID_DEVCLASS_BATTERY, idev, &did)) {
+      if (GetLastError() != ERROR_NO_MORE_ITEMS) {

```

Comment:
I'm not quite sure what this does. Does it cause an early return if the OS says we're done with batteries, or does it always try 100 times? (Because if it's the latter, I'm wondering if there's a speed/overhead impact)

Reply:
The intent is to stop enumeration if the OS reports we are out of batteries, and continue attempting if we get other errors (which I have not yet seen in my testing). I realize we are missing a `continue` after line 91.

Reply:
It does an early return no matter what the error is, but only logs if the error is something besides the OS saying there are no more batteries to enumerate (which is an expected case).